src/aten/sycl/Indexing.cpp

```diff
@@ -7,17 +7,160 @@
 #include <ATen/ATen.h>
 #include <ATen/AccumulateType.h>
 #include <ATen/Dispatch.h>
-#include <ATen/native/CanUse32BitIndexMath.h>
+#include <ATen/MemoryOverlap.h>
+#include <ATen/native/Resize.h>
 #include <ATen/native/TensorIterator.h>
+#include <aten/sycl/Indexing.h>
 #include <aten/sycl/Loops.h>
-#include <aten/sycl/TensorInfo.h>
-#include <comm/SYCLContext.h>
 
-namespace at {
-namespace native {
-namespace xpu {
+#include <comm/SYCLContext.h>
+#include <comm/TensorInfo.h>
 
 using namespace at::xpu::detail;
+using namespace at::xpu;
+
+namespace at::native::xpu {
+
+template <typename ValType>
+class IndexSelectScalarFunctor {
+ public:
+  void operator()(
+      ValType* dst,
+      ValType* src,
+      int64_t dst_off,
+      int64_t src_off,
+      int64_t idx,
+      ValType alpha) const {
+    dst[dst_off] = src[src_off];
+  }
+};
+
+template <
+    class SrcInfo,
+    class DstInfo,
+    class IdxInfo,
+    bool TrivialOffCal = false>
+static inline void _index_select_kernel(
+    SrcInfo& src_info,
+    DstInfo& dst_info,
+    IdxInfo& index_info,
+    int64_t dim) {
+  using scalar_t = typename SrcInfo::scalar_t;
+  auto cfg = IndexKernelConfig<
+      SrcInfo,
+      DstInfo,
+      IdxInfo,
+      IndexSelectScalarFunctor<scalar_t>>::
+      make_config(
+          src_info,
+          dst_info,
+          index_info,
+          static_cast<scalar_t>(0),
+          dim,
+          false,
+          IndexSelectScalarFunctor<scalar_t>());
+  if (cfg.problem_inner_) {
+    launch_index_kernel<decltype(cfg), TrivialOffCal, true>(cfg);
+  } else {
+    launch_index_kernel<decltype(cfg), TrivialOffCal, false>(cfg);
+  }
+}
+
+void index_select_kernel(
+    const Tensor& src,
+    int64_t dim,
+    const Tensor& indices,
+    const Tensor& dst) {
+  at::assert_no_internal_overlap(dst);
+  at::assert_no_overlap(dst, src);
+  at::assert_no_overlap(dst, indices);
```

Comment:
Do these operation require XPU specific implementation and register for XPU?

Reply:
No, they don't. The utils check the overlap of tensors by checking tensor storage on host.